pupil_src/shared_modules/gaze_mapping/gazer_2d_v1x.py

```diff
@@ -23,78 +22,121 @@
     make_map_function,
 )
 
+_REFERENCE_FEATURE_COUNT = 2
+
+_MONOCULAR_FEATURE_COUNT = 2
+_MONOCULAR_PUPIL_NORM_POS = slice(0, 2)
+
+_BINOCULAR_FEATURE_COUNT = 4
+_BINOCULAR_PUPIL_NORM_POS = slice(2, 4)
+
 
 class Model2D_v1x(Model):
-    def __init__(self, *, binocular, screen_size=(1, 1)):
-        self.binocular = binocular
+    def __init__(self, *, screen_size=(1, 1)):
         self.screen_size = screen_size
+        self._is_fitted = False
+
+    @property
+    def is_fitted(self) -> bool:
+        return self._is_fitted
+
+    def set_params(self, params, map_fn=None):
+        self._map_fn = map_fn or make_map_function(*params)
+        self._params = params
+
+    def get_params(self):
+        return {"params": self._params}
+
+
+class Model2D_v1x_Binocular(Model2D_v1x):
+    def fit(self, X, Y):
+        assert X.ndim == Y.ndim == 2, "Required shape: (n_samples, n_features)"
+        assert X.shape[0] == Y.shape[0], "Requires same number of samples in X and Y"
+
+        point_cloud = np.hstack([X, Y])  # specific to calibrate_2d_polynomial
+        num_expecte_features = _BINOCULAR_FEATURE_COUNT + _REFERENCE_FEATURE_COUNT
+        assert point_cloud.shape[1] == num_expecte_features
 
-    def fit(self, point_cloud):  # NOTE: breaks api due to legacy reasons
         map_fn, inliers, params = calibrate_2d_polynomial(
-            point_cloud, self.screen_size, binocular=self.binocular
+            point_cloud, self.screen_size, binocular=True
         )
         if not inliers.any():
             raise NotEnoughDataError
 
         self.set_params(map_fn=map_fn, params=params)
+        self._is_fitted = True
 
     def predict(self, X):
-        return [self._predict_single(x) for x in X]
-
-    def _predict_single(self, x):
-        if self.binocular:
-            p0, p1 = x
-            return self._map_fn(p0["norm_pos"], p1["norm_pos"])
-        else:
-            return self._map_fn(x[0]["norm_pos"])
+        # X[:, _MONOCULAR_PUPIL_NORM_POS] -> norm_pos left
+        # X[:, _BINOCULAR_PUPIL_NORM_POS] -> norm_pos right
+        return [
+            self._map_fn(x[_MONOCULAR_PUPIL_NORM_POS], x[_BINOCULAR_PUPIL_NORM_POS])
```

Comment:
Depending on how large `X` is, calling `self._map_fn` on every iteration might be slow. One easy way to fix that (assuming `self._map_fn` is a pure function that does not mutate `self`) is to declare `map_fn = self._map_fn` above the `return` and use `map_fn` inside the list comprehension.

Reply:
We plan on building a map function that actually operates on the numpy array itself instead of iterating it. This is just a work around for the existing map function.